src/Grpc.AspNetCore.Web/Internal/Base64PipeReader.cs

```diff
@@ -0,0 +1,193 @@
+            var length = _currentDecodedBuffer.Slice(0, base64Position).Length;
+
+            // IMPORTANT - this logic assumes that SequencePosition will be multiples of 3.
+            // GetMaxEncodedToUtf8Length will round up to next multiple of 3, so more data
+            // could be advanced internally than expected if not called using a mulitple of 3.
+            //
+            // Because ReadAsync always returns data in multiples of 3, and gRPC only uses
+            // ReadResult.Buffer.End to advance, the position should always be a multiple
+            // of 3. The only time it won't be is at the end of a message, in which case
+            // rounding up is the correct thing to do to consume padding.
+            var resolvedPosition = _currentInnerBuffer.GetPosition(Base64.GetMaxEncodedToUtf8Length((int)length));
+            return resolvedPosition;
+        }
+
+        public override void CancelPendingRead()
+        {
+            _inner.CancelPendingRead();
+        }
+
+        public override void Complete(Exception? exception = null)
+        {
+            _inner.Complete(exception);
+        }
+
+        public async override ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        {
+            var innerResult = await _inner.ReadAsync(cancellationToken);
+            if (innerResult.Buffer.IsEmpty)
+            {
+                _currentDecodedBuffer = innerResult.Buffer;
+                _currentInnerBuffer = innerResult.Buffer;
+                return innerResult;
+            }
+
+            // Minimum valid base64 length is 4. Read until we have at least that much content
+            while (innerResult.Buffer.Length < 4)
+            {
+                if (innerResult.IsCompleted)
+                {
+                    // If the reader completes with less than 4 bytes then the base64 isn't valid..
+                    throw new InvalidOperationException("Unexpected end of data when reading base64 content.");
+                }
+
+                if (innerResult.IsCanceled)
+                {
+                    return new ReadResult(
+                        ReadOnlySequence<byte>.Empty,
+                        innerResult.IsCanceled,
+                        innerResult.IsCompleted);
+                }
+
+                // Attempt to get more data
+                _inner.AdvanceTo(innerResult.Buffer.Start, innerResult.Buffer.End);
+                innerResult = await _inner.ReadAsync(cancellationToken);
+            }
+
+            // Limit result to complete base64 segments (multiples of 4)
+            var buffer = innerResult.Buffer.Slice(0, (innerResult.Buffer.Length / 4) * 4);
+
+            // The content can contain multiple fragments of base64 content
+            // Check for padding, and limit returned data to one fragment at a time
+            var paddingIndex = PositionOf(buffer, (byte)'=');
```

Comment:
Is there a way you could use [ReadOnlySequence.TryReadTo](https://github.com/dotnet/runtime/blob/master/src/libraries/System.Memory/src/System/Buffers/SequenceReader.Search.cs#L189) instead? I see you're doing some logic to make sure the inner buffer's length is a multiple of 4, so it might be tricky. The nice thing about TryReadTo is it basically combines PositionOf and Slice into a single operation meaning it doesn't need to loop over the segments twice in the multi-segment case.

Reply:
Need to split on the last padding character (in 4th position), not the first.